From: Duncan Eastoe <duncan@graphiant.com>
Date: Tue, 2 Aug 2022 14:10:13 +0100
Subject: sysrepo: add SR_UNDERLYING_VALUE option

This option is provided for use with sr_get_items() to allow easy
retrieval of the underlying integer value of enums, rather than the
typical string representation.

An application could achieve this with a combination sr_get_data() and
then looking up the schema. However the intention of this option is to
provide a more convenient alternative, particularly in Rust applications
(where this would currently require unsafe code).
---
 src/common.c                 |  4 +--
 src/common.h                 |  2 +-
 src/sysrepo.c                |  6 ++--
 src/sysrepo_types.h          | 10 ++++--
 tests/files/test-module.yang | 12 +++++--
 tests/test_get.c             | 77 +++++++++++++++++++++++++++++++++++++++++++-
 6 files changed, 101 insertions(+), 10 deletions(-)

diff --git a/src/common.c b/src/common.c
index 750ce18..635ed10 100644
--- a/src/common.c
+++ b/src/common.c
@@ -4048,11 +4048,11 @@ sr_ev2api(sr_sub_event_t ev)
 sr_error_info_t *
 sr_val_ly2sr(const struct lyd_node *node, int with_origin, sr_val_t *sr_val)
 {
-    return sr_val_ly2sr_with_opts(node, SR_VAL_LY2SR_DEFAULT, sr_val);
+    return sr_val_ly2sr_with_opts(node, with_origin, SR_VAL_LY2SR_DEFAULT, sr_val);
 }
 
 sr_error_info_t *
-sr_val_ly2sr_with_opts(const struct lyd_node *node, sr_val_ly2sr_opts_t opts, sr_val_t *sr_val)
+sr_val_ly2sr_with_opts(const struct lyd_node *node, int with_origin, sr_val_ly2sr_opts_t opts, sr_val_t *sr_val)
 {
     sr_error_info_t *err_info = NULL;
     char *ptr, *origin = NULL;
diff --git a/src/common.h b/src/common.h
index a883d35..efc8cd0 100644
--- a/src/common.h
+++ b/src/common.h
@@ -243,7 +243,7 @@ typedef enum sr_val_ly2sr_opts_e {
  * @param[out] sr_val sysrepo value.
  * @return err_info, NULL on success.
  */
-sr_error_info_t *sr_val_ly2sr_with_opts(const struct lyd_node *node, sr_val_ly2sr_opts_t opts, sr_val_t *sr_val);
+sr_error_info_t *sr_val_ly2sr_with_opts(const struct lyd_node *node, int with_origin, sr_val_ly2sr_opts_t opts, sr_val_t *sr_val);
 
 /*
  * From sysrepo.c
diff --git a/src/sysrepo.c b/src/sysrepo.c
index cd3663e..82f2d42 100644
--- a/src/sysrepo.c
+++ b/src/sysrepo.c
@@ -2895,7 +2895,7 @@ sr_get_items(sr_session_ctx_t *session, const char *xpath, uint32_t timeout_ms,
     uint32_t i;
 
     SR_CHECK_ARG_APIRET(!session || !xpath || !values || !value_cnt ||
-            ((session->ds != SR_DS_OPERATIONAL) && (opts & SR_OPER_MASK)), session, err_info);
+            ((session->ds != SR_DS_OPERATIONAL) && ((opts & ~SR_UNDERLYING_VALUE) & SR_OPER_MASK)), session, err_info);
 
     if (!timeout_ms) {
         timeout_ms = SR_OPER_CB_TIMEOUT;
@@ -2955,7 +2955,9 @@ sr_get_items(sr_session_ctx_t *session, const char *xpath, uint32_t timeout_ms,
     }
 
     for (i = 0; i < set->count; ++i) {
-        if ((err_info = sr_val_ly2sr(set->dnodes[i], opts & SR_OPER_WITH_ORIGIN, (*values) + i))) {
+        if ((err_info = sr_val_ly2sr_with_opts(set->dnodes[i], opts & SR_OPER_WITH_ORIGIN,
+                (opts & SR_UNDERLYING_VALUE) ? SR_VAL_LY2SR_UNDERLYING_VALUE : SR_VAL_LY2SR_DEFAULT,
+                (*values) + i))) {
             goto cleanup;
         }
         ++(*value_cnt);
diff --git a/src/sysrepo_types.h b/src/sysrepo_types.h
index 9232e3c..1daadc5 100644
--- a/src/sysrepo_types.h
+++ b/src/sysrepo_types.h
@@ -307,8 +307,13 @@ typedef enum sr_get_oper_flag_e {
                                           one inherit the origin from parents. */
     SR_OPER_NO_POLL_CACHED = 0x20,   /**< Do not use cached oper data from operational poll subscriptions even if
                                           available. */
-    SR_OPER_NO_RUN_CACHED = 0x40     /**< Do not use connection running datastore cache data even if the connection
+    SR_OPER_NO_RUN_CACHED = 0x40,    /**< Do not use connection running datastore cache data even if the connection
                                           supports it, may prevent some dead locks. */
+    SR_UNDERLYING_VALUE = 0x20000,   /**< Returns the underlying value for nodes with an appropriate type.
+                                          This causes nodes of type SR_ENUM_T to instead be returned as SR_INT32_T,
+                                          with the underlying numeric representation of the enum value.
+                                          There is currently no effect on any other node type.
+                                          This option works on data in any datastore. */
 } sr_get_oper_flag_t;
 
 #define SR_OPER_MASK 0xFFFF          /**< Mask for all get oper data flags. */
@@ -317,9 +322,10 @@ typedef enum sr_get_oper_flag_e {
  * @brief Flags used to override default data get behavior.
  */
 typedef enum {
-    SR_GET_NO_FILTER = 0x010000      /**< Do not apply the filter and return the whole "base" data which the filter
+    SR_GET_NO_FILTER = 0x010000,      /**< Do not apply the filter and return the whole "base" data which the filter
                                           would normally be applied on. The filter is used only when deciding what data
                                           to retrieve from subscribers and similar optimization cases. */
+
 } sr_get_flag_t;
 
 /**
diff --git a/tests/files/test-module.yang b/tests/files/test-module.yang
index 55f5d1e..c8d85af 100644
--- a/tests/files/test-module.yang
+++ b/tests/files/test-module.yang
@@ -40,8 +40,16 @@ module test-module {
         enum "no" {
           value 2;
         }
-        enum "maybe" {
-          value 3;
+        enum "maybe";
+      }
+    }
+    leaf enum-union {
+      type union {
+        type enumtpdf;
+        type enumeration {
+          enum "d" {
+            value 4;
+          }
         }
       }
     }
diff --git a/tests/test_get.c b/tests/test_get.c
index dd6e4b9..c7b8701 100644
--- a/tests/test_get.c
+++ b/tests/test_get.c
@@ -49,6 +49,8 @@ setup(void **state)
         TESTS_SRC_DIR "/files/simple-aug.yang",
         TESTS_SRC_DIR "/files/defaults.yang",
         TESTS_SRC_DIR "/files/mod.yang",
+        TESTS_SRC_DIR "/files/referenced-data.yang",
+        TESTS_SRC_DIR "/files/test-module.yang",
         NULL
     };
 
@@ -85,17 +87,89 @@ teardown(void **state)
         "simple",
         "defaults",
         "mod",
+        "test-module",
+        "referenced-data",
         NULL
     };
 
     sr_remove_modules(st->conn, module_names, 0);
-
     sr_disconnect(st->conn);
     sr_disconnect(st->cconn);
     free(st);
     return 0;
 }
 
+/* TEST */
+static void
+test_get_enum(void **state)
+{
+    struct state *st = (struct state *)*state;
+    sr_val_t *vals = NULL;
+    size_t vals_cnt = 0;
+    int ret;
+
+    /* set enum which has an explicit value */
+    ret = sr_set_item_str(st->sess, "/test-module:main/enum", "yes", NULL, 0);
+    assert_int_equal(ret, SR_ERR_OK);
+    ret = sr_apply_changes(st->sess, 0);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    /* read it back as an enum (string) */
+    ret = sr_get_items(st->sess, "/test-module:main/enum", 0, SR_OPER_DEFAULT, &vals, &vals_cnt);
+    assert_int_equal(ret, SR_ERR_OK);
+    assert_int_equal(vals[0].type, SR_ENUM_T);
+    assert_string_equal(vals[0].data.enum_val, "yes");
+    sr_free_values(vals, vals_cnt);
+
+    /* read it back as the underlying value (int) */
+    vals = NULL;
+    vals_cnt = 0;
+    ret = sr_get_items(st->sess, "/test-module:main/enum", 0, SR_UNDERLYING_VALUE, &vals, &vals_cnt);
+    assert_int_equal(ret, SR_ERR_OK);
+    assert_int_equal(vals[0].type, SR_INT32_T);
+    assert_int_equal(vals[0].data.int32_val, 1);
+    sr_free_values(vals, vals_cnt);
+
+    /* set enum which does not have an explicit value */
+    ret = sr_set_item_str(st->sess, "/test-module:main/enum", "maybe", NULL, 0);
+    assert_int_equal(ret, SR_ERR_OK);
+    ret = sr_apply_changes(st->sess, 0);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    /* read it back as an enum (string) */
+    vals = NULL;
+    vals_cnt = 0;
+    ret = sr_get_items(st->sess, "/test-module:main/enum", 0, SR_OPER_DEFAULT, &vals, &vals_cnt);
+    assert_int_equal(ret, SR_ERR_OK);
+    assert_int_equal(vals[0].type, SR_ENUM_T);
+    assert_string_equal(vals[0].data.enum_val, "maybe");
+    sr_free_values(vals, vals_cnt);
+
+    /* read it back as the underlying value (int) */
+    vals = NULL;
+    vals_cnt = 0;
+    ret = sr_get_items(st->sess, "/test-module:main/enum", 0, SR_UNDERLYING_VALUE, &vals, &vals_cnt);
+    assert_int_equal(ret, SR_ERR_OK);
+    assert_int_equal(vals[0].type, SR_INT32_T);
+    assert_int_equal(vals[0].data.int32_val, 3);
+    sr_free_values(vals, vals_cnt);
+
+    /* set union of two enums */
+    ret = sr_set_item_str(st->sess, "/test-module:main/enum-union", "d", NULL, 0);
+    assert_int_equal(ret, SR_ERR_OK);
+    ret = sr_apply_changes(st->sess, 0);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    /* read it back as the underlying value (int) */
+    vals = NULL;
+    vals_cnt = 0;
+    ret = sr_get_items(st->sess, "/test-module:main/enum-union", 0, SR_UNDERLYING_VALUE, &vals, &vals_cnt);
+    assert_int_equal(ret, SR_ERR_OK);
+    assert_int_equal(vals[0].type, SR_INT32_T);
+    assert_int_equal(vals[0].data.int32_val, 4);
+    sr_free_values(vals, vals_cnt);
+}
+
 /* TEST */
 static void
 test_invalid(void **state)
@@ -774,6 +848,7 @@ int
 main(void)
 {
     const struct CMUnitTest tests[] = {
+        cmocka_unit_test(test_get_enum),
         cmocka_unit_test(test_invalid),
         cmocka_unit_test(test_cached_datastore),
         cmocka_unit_test(test_cached_thread),
